### ISSUE TITLE
Datashade with hover information and colorization like datashader with Bokeh

### ISSUE THREAD

**Issue by dwervin** [ioam/holoviews#1795]
Using Holoviews and datashade, is there a way to get the visualization like the nyc_taxi notebook examples (Bokeh/datashader) and hover information?

It looks like datashader is more robust with Bokeh, but the hover functionality is not quite right.  But, the hover funtionality works well in holoviews, but datashade seems to not be as robust.

Is there a way to configure datashade in Holoviews to display something similar to the stacking example and the ds.count_cat example in the nys_taxi notebook that uses Bokeh and datashader?

Thanks

**Comment by jbednar** [comment]
See https://github.com/bokeh/datashader/issues/430#issuecomment-320792479 .

For what it's worth, datashader is more robust with HoloViews than with Bokeh alone, because with Bokeh you have to either do all the coding yourself (which is unlikely to turn out well!), or use InteractiveImage (which is very limited and fails to work in many cases where you would expect it to work.  We've only found it practical to support datashader using holoviews, because datashader needs dynamic functionality that is very hard to coordinate and work with otherwise.

**Comment by philippjfr** [comment]
@dwervin Yes this is possible although we are still going to be looking up the API for this. Here is one way of doing it right now:

```python
import holoviews as hv
import geoviews as gv
import pandas as pd
import dask.dataframe as dd

from colorcet import cm
from bokeh.models import WMTSTileSource
from holoviews.streams import RangeXY, PlotSize
from holoviews.operation.datashader import aggregate, shade, dynspread
import datashader.transfer_functions as tf
import datashader as ds

hv.extension('bokeh')

# Load data
df = dd.read_parquet('./data/nyc_taxi.parq/').persist()

# Declare tile source, pickups and dropoffs
wmts = WMTSTileSource(url='https://maps.wikimedia.org/osm-intl/{Z}/{X}/{Y}@2x.png')
pickups = hv.Points(df, kdims=['pickup_x', 'pickup_y'], vdims=['passenger_count'])
dropoffs = hv.Points(df, kdims=['dropoff_x', 'dropoff_y'], vdims=['passenger_count'])

def nyc_shaded(x_range, y_range):
    """
    Returns a RGB Element of stacked pickups and dropoffs 
    """
    
    # Aggregate pickups and dropoffs
    agg_kwargs = dict(x_range=x_range, y_range=y_range, dynamic=False,
                      width=height, height=height, aggregator=ds.count('passenger_count'))
    picks = aggregate(pickups, **agg_kwargs) 
    drops = aggregate(dropoffs, **agg_kwargs)
    
    # Get hold of xarray arrays
    picks = picks.redim(pickup_x='x', pickup_y='y')
    drops = drops.redim(dropoff_x='x', dropoff_y='y')
    picks, drops = picks.data.passenger_count, drops.data.passenger_count
    
    # Shade regions with more dropoffs than pickups and vice versa then stack
    more_drops = tf.shade(drops.where(drops > picks), cmap=["darkblue", 'cornflowerblue'], how='log')
    more_picks = tf.shade(picks.where(picks > drops), cmap=["darkred", 'orangered'],  how='log')
    img = tf.stack(more_picks,more_drops)
    rgb = tf.dynspread(img, threshold=0.3, max_px=4)
    
    # Declare Element with stacked RGB data
    bounds = (x_range[0], y_range[0], x_range[1], y_range[1])
    return hv.RGB(shade.uint32_to_uint8(rgb.data), bounds=bounds)

def nyc_pickup_hover(x_range, y_range):
    """
    Compute low-resolution aggregate and wrap in QuadMesh to use as hover layer
    """
    picks = aggregate(pickups, width=20, height=20, x_range=x_range, y_range=y_range,
                      dynamic=False, aggregator=ds.count('passenger_count')) 
    return hv.QuadMesh(picks, vdims=['Pickups'])
```

And then display it in a new cell:

```python
%%opts RGB [width=600 height=400] QuadMesh [tools=['hover']] (hover_line_alpha=1 fill_alpha=0 line_color='black')
x_range, y_range = ((-8242000,-8210000), (4965000,4990000))
range_stream = RangeXY(x_range=x_range, y_range=y_range)
gv.WMTS(wmts) *\
hv.DynamicMap(nyc_shaded, streams=[range_stream]) *\
hv.DynamicMap(nyc_pickup_hover, streams=[range_stream])
```

The main annoyance is in getting a hold of the xarray DataArrays to do comparison on, particularly because we have to rename the coordinate dimensions first. I'll keep thinking about how to make this easier. The other limitation is that the hover layer can only contain one value dimension atm, while in future it should support displaying both pickup and dropoff counts for instance.

**Comment by dwervin** [comment]
I really appreciate that example.  

This clears up an error I came across when trying to do something similar.  Since I didn't redimension like you did, I was essentially trying to pass two sets of x,y points and it threw an error.

I also see that I might have been tying my own hands since I was only importing from holoviews and not importing datashader.

Are there any gotchas with trying to use widgets with this example?

Thanks

**Comment by jbednar** [comment]
@philippjfr, can you comment on the pros and cons of merging the drops and picks into a single RGB as you show above, versus having two separate Image elements and returning an Overlay of Images instead?  I originally wrote the merging approach before there was HoloViews support, and I'm wondering if we shouldn't let HoloViews and Bokeh do more of the work here (e.g. so that there can be legends and other niceties).

**Comment by philippjfr** [comment]
>can you comment on the pros and cons of merging the drops and picks into a single RGB as you show above, versus having two separate Image elements and returning an Overlay of Images instead?

Two issues here, first of all an Overlay of Images won't work nicely at all right now because Image has no support for transparency. Secondly it's not clear to me if overlaying two images is the same as using ``tf.stack``, which supports several compositing modes.

**Comment by jbednar** [comment]
Good point; the transparency issue needs addressing in Bokeh first.  Here, tf.stack is being used to overlay two disjoint images that share no non-transparent pixels in common, and so any method of composition should be the same as any other.  Sounds like overlaying two RGB objects would work for the moment, but two Image objects would be more natural once Bokeh supports that.  Is there an issue open on the Bokeh repo about that problem?

**Comment by philippjfr** [comment]
There's two related issues with transparency for the image glyph:

1) NaN values cannot be mapped to zero alpha (originally discussed in https://github.com/bokeh/bokeh/issues/5323 and https://github.com/bokeh/bokeh/pull/5344, which are now both closed)

2) Allowing colormappers which generate partially transparent colors (never discussed or supported in bokeh).

**Comment by jbednar** [comment]
Can you please file a Bokeh issue about that?  I discussed it in https://github.com/bokeh/bokeh/pull/5344 but that's not an actual issue.  I think you can summarize the issue better than I can.

**Comment by dwervin** [comment]
@philippjfr, I tried to follow your example, using my data, but I get the following error.
         raise TypeError("agg must be instance of DataArray")

I noticed that your example used dask, but I was using a pandas dataframe.  Could it be a difference with how dask dataframes work versus pandas that is causing my issue?  


I went ahead and tried loading my data using dask and it loaded fine, but I run into some difficulty when trying to convert my points to Mercator X,Y.  My original data uses KS State Plane South which I convert to Lon/lat using Stateplane package and then to Mercator using datashader utilities.  The difficulty using dask appears to be when trying to load the lon,lat array into a the dask dataframe.  I can't seem to get it to work.  I assume that the datashader utils will work when converting lon,lat to Mercator in a dask dataframe, but I haven't gotten to that point yet.

I suppose I could convert points in pandas, write out the csv with the Mercator points and then reload the new csv using dask, but I was hoping there was a better way without the intermediate steps.

**Comment by dwervin** [comment]
I reworked some things and got a little further, but now I see the following error when trying to run this line:
 **more_calls = tf.shade(calls.where(calls > meters), cmap=["darkblue", 'cornflowerblue'], how='log')**

ValueError: conflicting sizes for dimension 'x': length 800 on the data but length 400 on coordinate 'x'

**Comment by jbednar** [comment]
> There's two related issues with transparency for the image glyph:

Philipp's PR to allow transparency has now been accepted into Bokeh, so the next Bokeh release should make this feasible.

> I see the following error when trying to run this line:

The message suggests that if you look at ``calls`` and ``meters``, you'll find that they are two different-sized arrays, at least in the x dimension (and probably in y).  So your problem would be somewhere before the line you showed.

**Comment by dwervin** [comment]
They appeared to be the same size when I checked, but I'll dig a little more.  Maybe I'm not looking at it correctly.

Thanks

**Comment by dwervin** [comment]
I'm stumped.  I don't see that they are a different size.  Any ideas?

**Comment by jbednar** [comment]
Try just ``calls > meters`` at the prompt, which should give the same error, if I'm reading your code correctly.  If it does, there's no datashader involved anymore -- ``calls`` should be an xarray of a certain size, and ``meters`` should be an xarray of a certain size.  If you post the result of evaluating those three expressions(``calls``, ``meters``, ``calls>meters``)  in different notebook cells it should be clear what's going on.

**Comment by dwervin** [comment]
Here is what I get for each.  They look the same, but maybe the points for calls and meters are exactly the same and that's causing the problem?


**calls**
<xarray.DataArray 'type_count' (y: 400, x: 400)>
array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ..., 
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])
Coordinates:
  * y        (y) float64 4.416e+06 4.417e+06 4.418e+06 4.419e+06 4.42e+06 ...
  * x        (x) float64 -1.097e+07 -1.097e+07 -1.097e+07 -1.097e+07 ...

**meters**
<xarray.DataArray 'type_count' (y: 400, x: 400)>
array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ..., 
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])
Coordinates:
  * y        (y) float64 4.416e+06 4.417e+06 4.418e+06 4.419e+06 4.42e+06 ...
  * x        (x) float64 -1.097e+07 -1.097e+07 -1.097e+07 -1.097e+07 ...

**calls > meters**
<xarray.DataArray 'type_count' (y: 400, x: 400)>
array([[False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       ..., 
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False]], dtype=bool)
Coordinates:
  * y        (y) float64 4.416e+06 4.417e+06 4.418e+06 4.419e+06 4.42e+06 ...
  * x        (x) float64 -1.097e+07 -1.097e+07 -1.097e+07 -1.097e+07 ...

**Comment by dwervin** [comment]
**calls.where(calls > meters)**


<xarray.DataArray 'type_count' (y: 400, x: 400)>
array([[ nan,  nan,  nan, ...,  nan,  nan,  nan],
       [ nan,  nan,  nan, ...,  nan,  nan,  nan],
       [ nan,  nan,  nan, ...,  nan,  nan,  nan],
       ..., 
       [ nan,  nan,  nan, ...,  nan,  nan,  nan],
       [ nan,  nan,  nan, ...,  nan,  nan,  nan],
       [ nan,  nan,  nan, ...,  nan,  nan,  nan]])
Coordinates:
  * y        (y) float64 4.416e+06 4.417e+06 4.418e+06 4.419e+06 4.42e+06 ...
  * x        (x) float64 -1.097e+07 -1.097e+07 -1.097e+07 -1.097e+07 ...

**Comment by jbednar** [comment]
Thanks! That looks suspiciously like a bug, if only our failure to raise a useful exception.  I think I have enough to go on now.

**Comment by dwervin** [comment]
You're welcome.  

It appears that the meter coordinates and calls coordinates were exactly the same.  As a test, I shifted one set slightly and I was able to get a plot, although it wasn't quite what I wanted.

While the coordinates are the same, there are characteristics about the data that would be interesting to compare where one column has a value different than the other.  Maybe that's how I should approach this now that I understand the data set better.

Thanks

**Comment by jbednar** [comment]
That was perfect; with that information it was very simple to reproduce it:

![image](https://user-images.githubusercontent.com/1695496/29300730-f81341f2-813c-11e7-8b7a-7aeb45e6fb42.png)

And the bug was clear almost immediately; because I was apparently insane in [November 2016](https://github.com/bokeh/datashader/commit/9843c77c), I typed ``.view(np.uint32)`` (which reinterprets bytes, making a double array appear to have twice as many values in each dimension as it really does) instead of ``.astype(np.uint32)`` (which actually converts the data).  In some cases that must have worked anyway, but I'll make a PR with the correct version.

In any case, this issue only occurs when there is no data in the image at all, so if you're getting the error, your plot would not be very useful.  Still, it shouldn't be an error!

**Comment by jbednar** [comment]
Fixed by #439.
